drivers/gpio/gpio_altera_pio.c

```diff
@@ -0,0 +1,381 @@
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_altera_configure(const struct device *dev,
+				 gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	const int port_pin_mask = cfg->common.port_pin_mask;
+	const int direction = cfg->direction;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	/* Check if pin number is within range */
+	if ((port_pin_mask & BIT(pin)) == 0) {
+		return -EINVAL;
+	}
+
+	/* Check if the direction is Bidirectional */
+	if (direction != 0) {
+		return -EINVAL;
+	}
+
+	addr = reg_base + ALTERA_AVALON_PIO_DIRECTION_OFFSET;
+
+	if (flags == GPIO_INPUT) {
+		sys_clear_bits(addr, BIT(pin));
+	} else if (flags == GPIO_OUTPUT) {
+		sys_set_bits(addr, BIT(pin));
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Get physical level of all input pins in a port.
+ *
+ * Read the value from the port. A low physical level on the
+ * pin will be interpreted as value 0. A high physical level
+ * will be interpreted as value 1.
+ *
+ * @param dev	Pointer to the device tructure for the driver instance.
+ * @param value Pointer to a variable where pin values will be stored.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_altera_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	addr = reg_base + ALTERA_AVALON_PIO_DATA_OFFSET;
+
+	__ASSERT(value != NULL, "value is null pointer!");
+
+	if	(value == NULL)	{
+		return -EINVAL;
+	}
+
+	*value = sys_read32((addr));
+
+	return 0;
+}
+
+/**
+ * @brief Set physical level of selected output pins to high.
+ *
+ * Set the port to 1. It will only write if data direction is set to output.
+ *
+ * @param dev  Pointer to the device structure for the driver instance.
+ * @param mask Value indicating which pins will be modified.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_altera_port_set_bits_raw(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	const int port_pin_mask = cfg->common.port_pin_mask;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	if ((port_pin_mask & mask) == 0) {
+		return -EINVAL;
+	}
+
+	if (!gpio_pin_direction(dev, mask)) {
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_GPIO_ALTERA_PIO_OUTSET
+	addr = reg_base + ALTERA_AVALON_PIO_SET_BITS;
+	sys_write32(mask, addr);
+#else
+	addr = reg_base + ALTERA_AVALON_PIO_DATA_OFFSET;
+	sys_set_bits(addr, mask);
+#endif
+
+	return 0;
+}
+
+/**
+ * @brief Set physical level of selected output pins to low.
+ *
+ * Set the port to 0. It will only write if data direction is set to output.
+ *
+ * @param dev  Pointer to the device structure for the driver instance.
+ * @param mask Value indicating which pins will be modified.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
```

Comment:
Please remove doxygen that duplicates (and possibly conflicts with) public driver api doxygen.